[Finishes #187354252-implementing-profile-page-settings

hatfelicien opened the pull request:
## Purpose

This pull request aims to implement user profiles in the application to allow users to maintain their personal information.

## Changes Made

Implemented user profile creation upon user registration.
Added fields for personal information such as name, email, gender, birthdate, language
Added fields for billing address.
Implemented profile editing functionality.
Associated a timestamp with each user profile.
Documented the feature using Swagger.

## Testing Instructions

Register as a new user in the application.
Verify that upon registration, a user profile is created with the provided personal information.
Verify that you can edit your own profile but not another user's profile.
Ensure that a timestamp is associated with each profile creation.
Review the Swagger documentation to understand the profile endpoints and their usage.

## Related Issues

## Checklist

Please review the following checklist and make sure all tasks are complete before submitting:

- [x] Code follows the project's coding standards
- [x] Changes are covered by tests
- [x] Documentation is updated (if applicable)
- [x] All tests pass

JeanIrad committed:
[finishes #187354240] setting up docker for backend

[finishes #187354240] setting up docker for backend

hatfelicien committed:
[Finishes #187354252-implementing-profile-page-settings

hatfelicien committed:
[Finishes #187354252-implementing-profile-page-settings

niyontwali commented on src/controllers/profileController.ts:
```diff
@@ -0,0 +1,67 @@
+import { Request, Response } from 'express';
+import User from '../database/models/user';
+
+const getAllUserProfiles = async (req: Request, res: Response): Promise<void> => {
+  try {
+    // Find all user profiles
+    const users = await User.findAll();
+
+    // Return all user profiles
+    res.status(200).json(users);
+  } catch (error) {
+    // Handle errors
+    console.error('Error fetching user profiles:', error);
+    res.status(500).send('Internal Server Error');
+  }
+};
```
Remove this, because we already have getAllUsers

niyontwali commented on src/controllers/profileController.ts:
```diff
@@ -0,0 +1,67 @@
+
+  try {
+    // Find the user by ID
+    const user = await User.findByPk(userId);
+    if (!user) {
+      res.status(404).json({ message: 'User not found' });
+      return;
+    }
+
+    // Return user profile
+    res.status(200).json({
+      id: user.id,
+      firstName: user.firstName,
+      lastName: user.lastName,
+      email: user.email,
+    });
```
Here, we need to retrieve a single user by their ID along with their roles. We require all the data pertaining to the user, considering it's a user profile. Therefore, you should fetch all the data for the user as well as any related data from other tables/models but for now it is just the model of Role only related.

niyontwali commented on src/controllers/profileController.ts:
```diff
@@ -0,0 +1,67 @@
+    });
+  } catch (error) {
+    // Handle errors
+    console.error('Error fetching user profile:', error);
```
remove this console and use the logger. Please refer to other controllers your teammates worked on.

niyontwali commented on src/controllers/profileController.ts:
```diff
@@ -0,0 +1,67 @@
+  } catch (error) {
+    // Handle errors
+    console.error('Error updating user profile:', error);
+    res.status(500).send('Internal Server Error');
```
Use the re-usable function in th validations. Please refer to how your teammates did it

niyontwali commented on src/controllers/profileController.ts:
```diff
@@ -0,0 +1,67 @@
+
+const updateUserProfile = async (req: Request, res: Response): Promise<void> => {
+  const userId: string = req.params.userId;
+  const newData = req.body;
```
specify which can be updated, for now from the user model, and please refer to how updateUser was done.

niyontwali commented on src/controllers/profileController.ts:
```diff
@@ -0,0 +1,67 @@
+    res.status(200).json({ message: 'User profile updated successfully' });
+  } catch (error) {
+    // Handle errors
+    console.error('Error updating user profile:', error);
```
remove console and use logger

niyontwali commented on src/routes/profileRoutes.ts:
```diff
@@ -0,0 +1,13 @@
+import { getAllUserProfiles, getUserProfile, updateUserProfile } from '../controllers/profileController';
+
+const router = express.Router();
+router.get('/users/', getAllUserProfiles);
```
You will need to remove this since its controller function will also be removed.

niyontwali commented on src/server.ts:
```diff
@@ -1,11 +1,14 @@
 import swaggerJsDoc from 'swagger-jsdoc';
 import options from './docs/swaggerdocs';
 import routes from './routes';
+//import { validateCreateProfile, validateUpdateProfile } from './validations/profileValidation';
```
remove this line, why comment it?

niyontwali commented on src/server.ts:
```diff
@@ -1,11 +1,14 @@
 import options from './docs/swaggerdocs';
 import routes from './routes';
+//import { validateCreateProfile, validateUpdateProfile } from './validations/profileValidation';
+import profileRoutes from './routes/profileRoutes';
```
remove this please, because you already have the profile route define in the overall routes that is in `/routes/index.ts`

niyontwali commented on src/server.ts:
```diff
@@ -23,32 +26,49 @@ const swaggerSpec = swaggerJsDoc(options);
 app.use('/api/docs', swaggerUI.serve, swaggerUI.setup(swaggerSpec));
 
 app.use('/api', routes);
+app.use('/docs', swaggerUI.serve, swaggerUI.setup(swaggerSpec));
```
remove this please, why did you add it there, yet it is alread on line 26 of this file

niyontwali commented on src/server.ts:
```diff
@@ -23,32 +26,49 @@ const swaggerSpec = swaggerJsDoc(options);
-  logger.error('Testing logger');
-  res.send('Welcome at Mavericks E-commerce Website Apis');
+// Profile routes
+app.use('/profiles', profileRoutes);
```
Remove this please, because all of this is in the index.ts or the routes

niyontwali commented on src/server.ts:
```diff
@@ -23,32 +26,49 @@ const swaggerSpec = swaggerJsDoc(options);
-  res.send('Welcome at Mavericks E-commerce Website Apis');
+// Profile routes
+app.use('/profiles', profileRoutes);
+
+// Error handling middleware for profile routes
+app.use((err: any, req: Request, res: Response, next: NextFunction) => {
+  console.error(err.stack);
+  res.status(500).send('Internal Server Error');
 });
```
Why did you add this? You should remove everything you added in the server.ts, I don't think your task involves changing anything from this file

niyontwali reviewed (changes requested):
Well done! However, there are several areas where you need to improve. Please carefully consider the feedback and apply it accordingly.

niyontwali committed:
Merge pull request #27 from atlp-rwanda/187354240-ch-dockerize-bn

[finishes #187354240] setting up docker for backend

Favor-star committed:
Fixed some misbehaviour in README.md

niyontwali committed:
Merge pull request #39 from atlp-rwanda/fit-fix-readme-file

Fixed some misbehaviour in README.md

amin-leon committed:
bugfix: configure swagger documentation url and login by checking status and verified fields

niyontwali committed:
Merge pull request #43 from atlp-rwanda/bugfix-swaggerDoc-url-configurations

bugfix: configure swagger documentation url and login by checking status and verified fields.

JeanIrad committed:
middleware functions feature completed
	modified:   src/controllers/userController.ts
	modified:   src/database/models/user.ts
	modified:   src/helpers/token.generator.ts
	new file:   src/middlewares/authMiddlewares.ts
	modified:   src/routes/roleRoute.ts
	modified:   tsconfig.json
	modified:   src/middlewares/authMiddlewares.ts

niyontwali committed:
Merge pull request #41 from atlp-rwanda/187454249-ft-middleware-fns

187454249 ft middleware fns

hozayves committed:
ft-resend-verification-link

niyontwali committed:
Merge pull request #45 from atlp-rwanda/resend-verification-link

ft-resend-verification-link

hatfelicien committed:
Merge branch '#187354252-ft-profile-page-settings' of https://github.com/atlp-rwanda/e-commerce-mavericcks-bn into #187354252-ft-profile-page-settings
